contrib/pkg/labelobjects/labelobjects.go

```diff
@@ -0,0 +1,928 @@
+	}
+
+	return nil
+}
+
+func (l *labelMigrator) labelClusterDeprovisions() error {
+	all := &hivev1alpha1.ClusterDeprovisionRequestList{}
+
+	err := l.client.List(context.TODO(), all)
+	if err != nil {
+		l.clusterDeprovisionsFailureCount++
+		return errors.Wrap(err, "Failed listing ClusterDeprovisionRequests")
+	}
+
+	for _, item := range all.Items {
+		objLogger := l.newObjectLogger(item.Name, item.Namespace, item.Kind)
+
+		name, found := findControllerOwnerNameByKind(&item, "ClusterDeployment")
+		if !found {
+			objLogger.Error("Missing ClusterDeployment owner reference")
+			continue
+		}
+
+		item.Labels = k8slabels.AddLabel(item.Labels, constants.ClusterDeploymentNameLabel, name)
+
+		if l.updateObject(objLogger.WithField("labels", item.Labels), &item) {
+			l.clusterDeprovisionsSuccessCount++
+		} else {
+			l.clusterDeprovisionsFailureCount++
+		}
+	}
+
+	if l.clusterDeprovisionsFailureCount > 0 {
+		return fmt.Errorf("Failed updating %v object(s)", l.clusterDeprovisionsFailureCount)
+	}
+
+	return nil
+}
+
+func (l *labelMigrator) labelDNSZones() error {
+	all := &hivev1alpha1.DNSZoneList{}
+
+	err := l.client.List(context.TODO(), all)
+	if err != nil {
+		l.dnsZonesFailureCount++
+		return errors.Wrap(err, "Failed listing DNSZones")
+	}
+
+	for _, item := range all.Items {
+		objLogger := l.newObjectLogger(item.Name, item.Namespace, item.Kind)
+
+		name, found := findControllerOwnerNameByKind(&item, "ClusterDeployment")
+		if !found {
+			objLogger.Error("Missing ClusterDeployment owner reference")
+			continue
+		}
+
+		item.Labels = k8slabels.AddLabel(item.Labels, constants.ClusterDeploymentNameLabel, name)
+		item.Labels = k8slabels.AddLabel(item.Labels, constants.DNSZoneTypeLabel, constants.DNSZoneTypeChild)
+
+		if l.updateObject(objLogger.WithField("labels", item.Labels), &item) {
+			l.dnsZonesSuccessCount++
+		} else {
+			l.dnsZonesFailureCount++
+		}
+	}
+
+	if l.dnsZonesFailureCount > 0 {
+		return fmt.Errorf("Failed updating %v object(s)", l.dnsZonesFailureCount)
+	}
+
+	return nil
+}
+
+func (l *labelMigrator) labelMergedPullSecrets() error {
+	all := &corev1.SecretList{}
+
+	err := l.client.List(context.TODO(), all)
+	if err != nil {
+		l.mergedPullSecretsFailureCount++
+		return errors.Wrap(err, "Failed listing Secrets")
+	}
+
+	for _, item := range all.Items {
+		objLogger := l.newObjectLogger(item.Name, item.Namespace, "Secret")
+
+		if item.Type != corev1.SecretTypeDockerConfigJson {
+			// Not a pull secret, wrong type
+			continue
+		}
+
+		if _, ok := item.Data[corev1.DockerConfigJsonKey]; !ok {
+			// Not a pull secret, missing key
+			continue
+		}
+
+		name, found := findControllerOwnerNameByKind(&item, "ClusterDeployment")
+		if !found {
+			continue
+		}
+
+		item.Labels = k8slabels.AddLabel(item.Labels, constants.ClusterDeploymentNameLabel, name)
+		item.Labels = k8slabels.AddLabel(item.Labels, constants.SecretTypeLabel, constants.SecretTypeMergedPullSecret)
```

Comment:
This looks a little off, it seems odd to assume that any secret of type dockerconfigjson, with a ClusterDeployment owner, is a merged pull secret, but on closer inspection that does appear to be correct, normal non-merged pull secrets would not be owned by their ClusterDeployment. Just calling it out in case anyone sees a flaw in my logic.

Reply:
It is probably true for the Dedicated clusters. It is not necessarily true for all users of Hive. For example, I would not be surprised if MCM put an owner ref from the non-merged pull secret to the clusterdeployment so that the secret was cleaned up when the clusterdeployment was.

Reply:
Forget to mention in the comment that since it is true for Dedicated, then we are probably good. Unless we are expecting other users of v1alpha1 to use our migration steps to migrate to v1--which scares me.

Reply:
But it may be better to use the same pattern that is being used for the Hive-managed syncsets. Namely, we can look for the merged pull secret by the expected name of the secret.

Reply:
I agree. Thanks for the feedback.